resources/todomvc/architecture-examples/react-redux/src/components/text-input.jsx

```diff
@@ -0,0 +1,55 @@
+import { Component } from "react";
+import classnames from "classnames";
+import PropTypes from "prop-types";
+
+export default class TextInput extends Component {
+    static propTypes = {
+        onSave: PropTypes.func.isRequired,
+        text: PropTypes.string,
+        placeholder: PropTypes.string,
+        editing: PropTypes.bool,
+        newTodo: PropTypes.bool,
+    };
+
+    state = {
+        text: this.props.text || "",
+    };
+
+    handleSubmit = (e) => {
+        const text = e.target.value.trim();
+        if (e.key.match(/Enter/i)) {
+            this.props.onSave(text);
+            if (this.props.newTodo) {
+                this.setState({ text: "" });
+            }
+        }
+    };
+
+    handleChange = (e) => {
+        this.setState({ text: e.target.value });
+    };
+
+    handleBlur = (e) => {
+        if (!this.props.newTodo)
+            this.props.onSave(e.target.value);
+    };
+
+    render() {
+        return (
+            <input
+                className={classnames({
+                    edit: this.props.editing,
+                    "new-todo": this.props.newTodo,
+                })}
+                type="text"
+                data-testid="text-input"
+                placeholder={this.props.placeholder}
+                autoFocus
+                value={this.state.text}
```

Comment:
Interesting that we have a controlled component in this version :-) I'm not against it, it's interesting that we have both approaches in the various benchmarks.